apps/src/music/player/ToneJSPlayer.ts

```diff
@@ -0,0 +1,193 @@
+  instrument: string;
+  events: {notes: string[]; transportTime: string}[];
+}
+
+class ToneJSPlayer {
+  private samplers: {[instrument: string]: typeof Tone.Sampler};
+  private activePlayers: (typeof Tone.GrainPlayer)[];
+  private currentPreview: {id: string; player: typeof Tone.GrainPlayer} | null;
+
+  constructor(
+    bpm = DEFAULT_BPM,
+    private readonly soundCache: SoundCache = new SoundCache()
+  ) {
+    Tone.Transport.bpm.value = bpm;
+    this.activePlayers = [];
+    this.samplers = [];
+    this.currentPreview = null;
+  }
+
+  getCurrentPosition(): string {
+    return Tone.Transport.position;
+  }
+
+  goToPosition(position: string) {
+    Tone.Transport.position = position;
+  }
+
+  setLoopEnabled(enabled: boolean) {
+    Tone.Transport.loop = enabled;
+  }
+
+  setLoopStart(startPosition: string) {
+    Tone.Transport.loopStart = startPosition;
+  }
+
+  setLoopEnd(endPosition: string) {
+    Tone.Transport.loopEnd = endPosition;
+  }
+
+  async loadSounds(sampleIds: string[], callbacks?: SoundLoadCallbacks) {
+    return this.soundCache.loadSounds(sampleIds, callbacks);
+  }
+
+  async loadInstrument(
+    instrumentName: string,
+    sampleMap: {[note: number]: string},
+    callbacks?: SoundLoadCallbacks
+  ) {
+    const urls: {[note: number]: AudioBuffer} = {};
+    await this.soundCache.loadSounds(Object.values(sampleMap), callbacks);
+    Object.keys(sampleMap).forEach(note => {
+      const buffer = this.soundCache.getSound(sampleMap[parseInt(note)]);
+      if (buffer) {
+        urls[parseInt(note)] = buffer;
+      }
+    });
+
+    const sampler = new Tone.Sampler(urls).toDestination();
+    this.samplers[instrumentName] = sampler;
+  }
+
+  async playSampleImmediately(sound: SampleEvent, onStop?: () => void) {
+    if (this.currentPreview) {
+      this.currentPreview.player.stop();
+    }
+
+    const buffer = await this.soundCache.loadSound(sound.id);
+    if (!buffer) {
+      console.log('not in cache');
+      return;
+    }
+
+    const playbackRate = Tone.Transport.bpm.value / sound.originalBpm;
+
+    const player = new Tone.GrainPlayer({
+      url: buffer,
+      grainSize: playbackRate * 0.1,
+    }).toDestination();
+
+    player.onstop = () => {
+      console.log('stopping ' + sound.id);
+      player.dispose();
+
+      if (this.currentPreview?.id === sound.id) {
+        this.currentPreview = null;
+      }
+
+      onStop?.();
+    };
+
+    player.detune = sound.pitchShift * 100;
+    player.playbackRate = playbackRate;
+    player.start();
+
+    this.currentPreview = {id: sound.id, player};
+  }
+
+  playSequenceImmediately(sequence: SamplerSequence) {
+    sequence.events.forEach(({notes, transportTime}) => {
+      this.samplers[sequence.instrument].triggerAttack(
+        notes,
+        `+${Tone.Transport.toSeconds(transportTime)}`
+      );
+    });
+  }
+
+  cancelPreviews() {
+    if (this.currentPreview) {
+      this.currentPreview.player.stop();
+    }
```

Comment:
Instead of calling this.currentPreview, consider wrapping it around a function that will wrap any disposing operations. 

Even if you decide to ignore the memory leak suggestion, it can still help you adding more features later like logging when issues happening here. 

```
  private disposeCurrentPreview() {
    if (this.currentPreview) {
      this.currentPreview.player.stop();
      this.currentPreview.player.dispose();
      this.currentPreview = null;
    }
  }
```